devito/at_controller.py

```diff
@@ -0,0 +1,211 @@
+import random
+from os import mkdir, path
+
+import logger
+
+
+class AutoTuner(object):
+
+    def __init__(self, op, blocked_dims=None, at_report_dir=None):
+        """Object responsible for auto tuning block sizes.
+
+        :param op: Operator object.
+        :param blocked_dims: list indicating which dims are blocked and we want to tune
+                             Default all except inner most.
+        :param at_report_dir: string - indicating path to auto tuning report directory.
+                              If not set current directory is used
+        :param blocked_dims: list indicating which dimensions we want to auto tune
+        :raises ValueError: if operator is not of Operator type
+        """
+
+        from operator import Operator
+        if not isinstance(op, Operator):
+            raise ValueError("AT requires Operator object to be passed as an argument")
+
+        self.op = op
+        self.op.propagator.profile = True
+
+        default_blocked_dims = ([True] * (len(self.op.shape) - 1))
+        default_blocked_dims.append(False)  # By default we dont auto tune inner most dim
+        self.blocked_dims = blocked_dims or default_blocked_dims
+
+        default_at_dir = path.join(path.dirname(path.realpath(__file__)), "At Report")
+        self.report_dir = None
+        self.final_report_path = None
+        self.set_report_path(at_report_dir or default_at_dir)
+        self.model_desc_template = "%s %d %d %s %s"
+
+        if not path.isdir(self.report_dir):  # Creates report dir if does not exist
+            mkdir(self.report_dir)
+
+    def set_report_path(self, report_path):
```

Comment:
This little utility routine seems superfluous. Why not do this all in-place above?